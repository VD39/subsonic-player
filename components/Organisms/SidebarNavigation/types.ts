export interface SidebarButtonProps {
  icon: string;
  text: string;
}
