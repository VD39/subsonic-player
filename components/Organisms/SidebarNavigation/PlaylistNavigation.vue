<script setup lang="ts">
import ButtonLink from '@/components/Atoms/ButtonLink.vue';

import NavigationItem from './Items/NavigationItem.vue';
import SubNavigationItem from './Items/SubNavigationItem.vue';

defineProps<{
  collapsed: boolean;
  playlists: Playlist[];
}>();

defineEmits<{
  addPlaylist: [];
}>();
</script>

<template>
  <ul class="mBM">
    <NavigationItem :collapsed="collapsed" title="Playlist">
      <template v-if="playlists.length">
        <SubNavigationItem
          v-for="playlist in playlists.slice(0, 5)"
          :key="playlist.id"
          :collapsed="collapsed"
          :icon="ICONS.playlist"
          :title="playlist.name"
          :to="`/playlist/${playlist.id}`"
        />
      </template>
    </NavigationItem>

    <li>
      <ButtonLink
        ref="addPlaylist"
        class="sidebarLink"
        full-width
        :icon="ICONS.playlistAdd"
        :show-text="!collapsed"
        title="Add playlist"
        @click="$emit('addPlaylist')"
      >
        Add playlist
      </ButtonLink>
    </li>
  </ul>
</template>
