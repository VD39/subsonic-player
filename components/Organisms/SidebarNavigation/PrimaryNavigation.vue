<script setup lang="ts">
import NavigationItem from './Items/NavigationItem.vue';
import SubNavigationItem from './Items/SubNavigationItem.vue';

defineProps<{
  collapsed: boolean;
}>();

const { drop } = useDragAndDrop();
</script>

<template>
  <ul class="mBM">
    <NavigationItem
      v-for="items in SIDEBAR_DESKTOP_NAVIGATION"
      :key="`navigation-${items.title}`"
      class="mBM"
      :collapsed
      :title="items.title"
    >
      <SubNavigationItem
        v-for="item in items.items"
        :key="`item-${item.title}`"
        :class="[
          {
            [DRAG_AND_DROP_CLASS_NAMES.isDroppable]: item.isDroppable,
          },
        ]"
        :collapsed
        :icon="item.icon"
        :title="item.title"
        :to="item.to"
        @drop="drop(QUEUE_ID, $event)"
      />
    </NavigationItem>
  </ul>
</template>
