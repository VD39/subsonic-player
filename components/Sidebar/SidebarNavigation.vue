<script setup lang="ts">
import SidebarItems from '@/components/Sidebar/SidebarItems.vue';
import IconButton from '@/components/Buttons/IconButton.vue';
import MusicLogo from '@/components/Logo/MusicLogo.vue';

const { collapsed, toggle } = useSidebar();

const buttonProps = computed<ButtonProps>(() =>
  collapsed.value
    ? {
        icon: 'PhTextIndent',
        text: 'Open Navigation',
      }
    : {
        icon: 'PhTextOutdent',
        text: 'Close Navigation',
      },
);
</script>

<template>
  <div :class="$style.sideBar">
    <div :class="$style.header">
      <MusicLogo :class="$style.logo" />

      <IconButton
        ref="toggleButton"
        :icon="buttonProps.icon"
        :title="buttonProps.text"
        @click="toggle"
      >
        {{ buttonProps.text }}
      </IconButton>
    </div>

    <nav :class="$style.navigationWrapper">
      <ul :class="$style.navigation">
        <SidebarItems
          v-for="(item, index) in SIDEBAR_NAVIGATION"
          :key="`navigation-${index}`"
          :title="item.title"
          :title-icon="item.icon"
          :title-to="item.to"
          :items="item.items"
        />
      </ul>
    </nav>
  </div>
</template>

<style module>
.sideBar {
  @mixin align-center;

  position: fixed;
  inset: auto 0 0;
  z-index: 1;
  width: 100%;
  height: var(--header-height);
  padding-top: 0;
  background-color: var(--background-color);
  border-top: 1px solid var(--border-color);
  transition: width var(--transition);

  @media (--tablet-up) {
    inset: 0 auto 0 0;
    display: unset;
    width: var(--sidebar-width);
    height: auto;
    min-height: 100vh;
    overflow-x: auto;
    border-top: none;
    border-right: 1px solid var(--border-color);
  }
}

.header {
  display: none;

  @media (--tablet-up) {
    @mixin align-center;

    justify-content: space-between;
    min-height: var(--header-height);
    padding: 0 var(--space-16) 0 var(--space-24);
    margin-bottom: var(--space-16);

    :global(.collapsed) & {
      justify-content: center;
      padding: 0 var(--space-24);
    }
  }
}

.logo {
  :global(.collapsed) & {
    @mixin visually-hidden;
  }
}

.navigationWrapper {
  width: 100%;
}

.navigation {
  @media (--mobile-only) {
    @mixin inner;

    display: flex;
    gap: var(--space-40);
    justify-content: space-between;
    font-size: var(--small-font-size);
  }
}
</style>
