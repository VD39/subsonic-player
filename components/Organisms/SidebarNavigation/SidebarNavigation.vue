<script setup lang="ts">
import ButtonLink from '@/components/Atoms/ButtonLink.vue';
import MusicLogo from '@/components/Molecules/MusicLogo.vue';

import PlaylistNavigation from './PlaylistNavigation.vue';
import PrimaryNavigation from './PrimaryNavigation.vue';

const { collapsed, toggle } = useSidebar();

const buttonProps = computed<ButtonProps>(() => ({
  icon: collapsed.value ? ICONS.sidebarCollapsed : ICONS.sidebarNotCollapsed,
  text: collapsed.value ? 'Open Navigation' : 'Close Navigation',
}));
</script>

<template>
  <div
    :class="[
      'centerItems',
      'column',
      $style.sidebarNavigation,
      {
        [$style.collapsed]: collapsed,
      },
    ]"
  >
    <div :class="['mBM', 'centerAll', 'spaceBetween', $style.header]">
      <MusicLogo :class="$style.musicLogo" />

      <ButtonLink
        ref="toggleButton"
        :icon="buttonProps.icon"
        :title="buttonProps.text"
        @click="toggle"
      >
        {{ buttonProps.text }}
      </ButtonLink>
    </div>

    <nav class="fullWidth">
      <PrimaryNavigation :collapsed="collapsed" />

      <PlaylistNavigation :collapsed="collapsed" />
    </nav>
  </div>
</template>

<style module>
.sidebarNavigation {
  position: fixed;
  inset: 0 auto var(--sidebar-bottom) 0;
  z-index: 9;
  width: var(--sidebar-width);
  padding-top: 0;
  overflow-x: auto;
  background-color: var(--background-color);
  border-right: 1px solid var(--border-color);
  transition: width var(--transition);
}

.header {
  --header-padding: 0 var(--space-16) 0 var(--space-24);

  min-height: var(--header-height);
  padding: var(--header-padding);

  .collapsed & {
    --header-padding: 0 var(--space-24);
  }
}

.musicLogo {
  .collapsed & {
    display: none;
  }
}
</style>
