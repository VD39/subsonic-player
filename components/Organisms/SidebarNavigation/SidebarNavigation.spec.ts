import type { VueWrapper } from '@vue/test-utils';

import { mockNuxtImport } from '@nuxt/test-utils/runtime';
import { mount } from '@vue/test-utils';

import SidebarNavigation from './SidebarNavigation.vue';

const toggleMock = vi.fn();
const collapsedMock = ref(false);

mockNuxtImport('useSidebar', () => () => ({
  collapsed: collapsedMock,
  toggle: toggleMock,
}));

function factory(props = {}) {
  return mount(SidebarNavigation, {
    props: {
      ...props,
    },
  });
}

describe('SidebarNavigation', () => {
  let wrapper: VueWrapper;

  beforeEach(() => {
    wrapper = factory();
  });

  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('when collapsed is false', () => {
    it('does not add the collapsed class to the wrapper element', () => {
      expect(wrapper.classes()).not.toContain('collapsed');
    });

    it('sets the correct icon on the toggle button', () => {
      expect(wrapper.findComponent({ ref: 'toggleButton' }).props('icon')).toBe(
        'PhTextOutdent',
      );
    });

    it('sets the correct title attribute on the toggle button', () => {
      expect(
        wrapper.findComponent({ ref: 'toggleButton' }).attributes('title'),
      ).toBe('Close Navigation');
    });

    it('sets the correct slot data on the toggle button', () => {
      expect(wrapper.findComponent({ ref: 'toggleButton' }).text()).toContain(
        'Close Navigation',
      );
    });
  });

  describe('when collapsed is true', () => {
    beforeEach(() => {
      collapsedMock.value = true;
      wrapper = factory();
    });

    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('adds the collapsed class to the wrapper element', () => {
      expect(wrapper.classes()).toContain('collapsed');
    });

    it('sets the correct icon on the toggle button', () => {
      expect(wrapper.findComponent({ ref: 'toggleButton' }).props('icon')).toBe(
        'PhTextIndent',
      );
    });

    it('sets the correct title attribute on the toggle button', () => {
      expect(
        wrapper.findComponent({ ref: 'toggleButton' }).attributes('title'),
      ).toBe('Open Navigation');
    });

    it('sets the correct slot data on the toggle button', () => {
      expect(wrapper.findComponent({ ref: 'toggleButton' }).text()).toContain(
        'Open Navigation',
      );
    });
  });

  describe('when the toggle ButtonLink component is clicked', () => {
    beforeEach(() => {
      wrapper.findComponent({ ref: 'toggleButton' }).vm.$emit('click');
    });

    it('calls the toggle function', () => {
      expect(toggleMock).toHaveBeenCalled();
    });
  });
});
