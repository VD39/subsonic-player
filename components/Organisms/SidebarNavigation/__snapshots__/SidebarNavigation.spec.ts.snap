// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SidebarNavigation > matches the snapshot 1`] = `
"<div class="centerItems column sidebarNavigation">
  <div class="mBM centerAll spaceBetween header"><a class="centerItems buttonLink musicLogo musicLogo" title="Home">
      <phwavesine class="icon" aria-hidden="true" size="32" weight="regular"></phwavesine><span class="clamp text visuallyHidden"> Home </span>
    </a><button class="centerItems buttonLink" type="button" title="Close Navigation">
      <phtextoutdent class="icon" aria-hidden="true" size="24" weight="regular"></phtextoutdent><span class="clamp text visuallyHidden">Close Navigation</span>
    </button></div>
  <nav class="fullWidth">
    <ul class="mBM">
      <li class="mBM">
        <h4 class="strong mBM sidebarTitle">Browse</h4>
        <ul>
          <li><a exactactiveclass="sidebarCurrentLink" class="centerItems buttonLink sidebarLink sidebarLink" title="Discover">
              <phcompass class="icon" aria-hidden="true" size="24" weight="regular"></phcompass><span class="clamp text">Discover</span>
            </a></li>
          <li><a exactactiveclass="sidebarCurrentLink" class="centerItems buttonLink sidebarLink sidebarLink" title="Podcasts">
              <phapplepodcastslogo class="icon" aria-hidden="true" size="24" weight="regular"></phapplepodcastslogo><span class="clamp text">Podcasts</span>
            </a></li>
          <li><a exactactiveclass="sidebarCurrentLink" class="centerItems buttonLink sidebarLink sidebarLink" title="Radio Stations">
              <phradio class="icon" aria-hidden="true" size="24" weight="regular"></phradio><span class="clamp text">Radio Stations</span>
            </a></li>
        </ul>
      </li>
      <li class="mBM">
        <h4 class="strong mBM sidebarTitle">Your library</h4>
        <ul>
          <li><a exactactiveclass="sidebarCurrentLink" class="centerItems buttonLink sidebarLink sidebarLink" title="Albums">
              <phvinylrecord class="icon" aria-hidden="true" size="24" weight="regular"></phvinylrecord><span class="clamp text">Albums</span>
            </a></li>
          <li><a exactactiveclass="sidebarCurrentLink" class="centerItems buttonLink sidebarLink sidebarLink" title="Artists">
              <phusersthree class="icon" aria-hidden="true" size="24" weight="regular"></phusersthree><span class="clamp text">Artists</span>
            </a></li>
          <li><a exactactiveclass="sidebarCurrentLink" class="centerItems buttonLink sidebarLink sidebarLink" title="Playlists">
              <phplaylist class="icon" aria-hidden="true" size="24" weight="regular"></phplaylist><span class="clamp text">Playlists</span>
            </a></li>
          <li><a exactactiveclass="sidebarCurrentLink" class="centerItems buttonLink sidebarLink sidebarLink" title="Favourites">
              <phheart class="icon" aria-hidden="true" size="24" weight="regular"></phheart><span class="clamp text">Favourites</span>
            </a></li>
          <li><a exactactiveclass="sidebarCurrentLink" class="centerItems buttonLink sidebarLink sidebarLink" title="Genres">
              <phwaveform class="icon" aria-hidden="true" size="24" weight="regular"></phwaveform><span class="clamp text">Genres</span>
            </a></li>
        </ul>
      </li>
    </ul>
    <ul class="mBM">
      <li>
        <h4 class="strong mBM sidebarTitle">Playlist</h4>
        <ul></ul>
      </li>
      <li><button class="centerItems buttonLink centerAll fullWidth sidebarLink sidebarLink" type="button" title="Add playlist">
          <phlistplus class="icon" aria-hidden="true" size="24" weight="regular"></phlistplus><span class="clamp text"> Add playlist </span>
        </button></li>
    </ul>
  </nav>
</div>"
`;

exports[`SidebarNavigation > when collapsed is true > matches the snapshot 1`] = `
"<div class="centerItems column sidebarNavigation collapsed">
  <div class="mBM centerAll spaceBetween header"><a class="centerItems buttonLink musicLogo musicLogo" title="Home">
      <phwavesine class="icon" aria-hidden="true" size="32" weight="regular"></phwavesine><span class="clamp text visuallyHidden"> Home </span>
    </a><button class="centerItems buttonLink" type="button" title="Open Navigation">
      <phtextindent class="icon" aria-hidden="true" size="24" weight="regular"></phtextindent><span class="clamp text visuallyHidden">Open Navigation</span>
    </button></div>
  <nav class="fullWidth">
    <ul class="mBM">
      <li class="mBM">
        <h4 class="strong mBM sidebarTitle visuallyHidden">Browse</h4>
        <ul>
          <li><a exactactiveclass="sidebarCurrentLink" class="centerItems buttonLink sidebarLink sidebarLink" title="Discover">
              <phcompass class="icon" aria-hidden="true" size="24" weight="regular"></phcompass><span class="clamp text visuallyHidden">Discover</span>
            </a></li>
          <li><a exactactiveclass="sidebarCurrentLink" class="centerItems buttonLink sidebarLink sidebarLink" title="Podcasts">
              <phapplepodcastslogo class="icon" aria-hidden="true" size="24" weight="regular"></phapplepodcastslogo><span class="clamp text visuallyHidden">Podcasts</span>
            </a></li>
          <li><a exactactiveclass="sidebarCurrentLink" class="centerItems buttonLink sidebarLink sidebarLink" title="Radio Stations">
              <phradio class="icon" aria-hidden="true" size="24" weight="regular"></phradio><span class="clamp text visuallyHidden">Radio Stations</span>
            </a></li>
        </ul>
      </li>
      <li class="mBM">
        <h4 class="strong mBM sidebarTitle visuallyHidden">Your library</h4>
        <ul>
          <li><a exactactiveclass="sidebarCurrentLink" class="centerItems buttonLink sidebarLink sidebarLink" title="Albums">
              <phvinylrecord class="icon" aria-hidden="true" size="24" weight="regular"></phvinylrecord><span class="clamp text visuallyHidden">Albums</span>
            </a></li>
          <li><a exactactiveclass="sidebarCurrentLink" class="centerItems buttonLink sidebarLink sidebarLink" title="Artists">
              <phusersthree class="icon" aria-hidden="true" size="24" weight="regular"></phusersthree><span class="clamp text visuallyHidden">Artists</span>
            </a></li>
          <li><a exactactiveclass="sidebarCurrentLink" class="centerItems buttonLink sidebarLink sidebarLink" title="Playlists">
              <phplaylist class="icon" aria-hidden="true" size="24" weight="regular"></phplaylist><span class="clamp text visuallyHidden">Playlists</span>
            </a></li>
          <li><a exactactiveclass="sidebarCurrentLink" class="centerItems buttonLink sidebarLink sidebarLink" title="Favourites">
              <phheart class="icon" aria-hidden="true" size="24" weight="regular"></phheart><span class="clamp text visuallyHidden">Favourites</span>
            </a></li>
          <li><a exactactiveclass="sidebarCurrentLink" class="centerItems buttonLink sidebarLink sidebarLink" title="Genres">
              <phwaveform class="icon" aria-hidden="true" size="24" weight="regular"></phwaveform><span class="clamp text visuallyHidden">Genres</span>
            </a></li>
        </ul>
      </li>
    </ul>
    <ul class="mBM">
      <li>
        <h4 class="strong mBM sidebarTitle visuallyHidden">Playlist</h4>
        <ul></ul>
      </li>
      <li><button class="centerItems buttonLink centerAll fullWidth sidebarLink sidebarLink" type="button" title="Add playlist">
          <phlistplus class="icon" aria-hidden="true" size="24" weight="regular"></phlistplus><span class="clamp text visuallyHidden"> Add playlist </span>
        </button></li>
    </ul>
  </nav>
</div>"
`;
